src/wms/src/main/java/org/geoserver/wms/map/RenderedImageMapOutputFormat.java

```diff
@@ -505,57 +509,86 @@ public RenderedImageMap produceMap(final WMSMapContent mapContent, final boolean
         
         onBeforeRender(renderer);
 
-        // setup the timeout enforcer (the enforcer is neutral when the timeout is 0)
+        //Timeout on the smallest nonzero value of the WMS timeout and the timeout format option
+        //If both are zero then there is no timeout
         int maxRenderingTime = wms.getMaxRenderingTime() * 1000;
-        RenderingTimeoutEnforcer timeout = new RenderingTimeoutEnforcer(maxRenderingTime, renderer,
-                graphic);
-        timeout.start();
-        try {
-            // finally render the image;
-            renderer.paint(graphic, paintArea, mapContent.getRenderingArea(),
-                    mapContent.getRenderingTransform());
-
-            // apply watermarking
-            if (layout != null) {
-                try {
-                    layout.paint(graphic, paintArea, mapContent);
-                } catch (Exception e) {
-                    throw new ServiceException("Problem occurred while trying to watermark data", e);
-                }
+        int localMaxRenderingTime = 0;
+        
+        Object timeoutOption = request.getFormatOptions().get("timeout");
+        if (timeoutOption != null) {
+            try {
+                localMaxRenderingTime = Integer.parseInt(timeoutOption.toString());
+            } catch (NumberFormatException e) {
+                LOGGER.log(Level.WARNING,"Could not parse format_option \"timeout\": "+timeoutOption, e);
             }
-        } finally {
-            timeout.stop();
-            graphic.dispose();
         }
-
-        // check if the request did timeout
-        if (timeout.isTimedOut()) {
-            throw new ServiceException(
-                    "This requested used more time than allowed and has been forcefully stopped. "
-                            + "Max rendering time is " + (maxRenderingTime / 1000.0) + "s");
+        if (maxRenderingTime == 0) {
+            maxRenderingTime = localMaxRenderingTime;
+        } else if (localMaxRenderingTime != 0) {
+            maxRenderingTime = Math.min(maxRenderingTime, localMaxRenderingTime);
         }
-
-        // check if a non ignorable error occurred
-        if (nonIgnorableExceptionListener.exceptionOccurred()) {
-            Exception renderError = nonIgnorableExceptionListener.getException();
-            throw new ServiceException("Rendering process failed", renderError, "internalError");
+        
+        //Run rendering in a separate thread so that we can exit promptly if the timeout is reached,
+        //rather than waiting for the renderer to finish the current task.
+        MapPainter painter = new MapPainter(renderer, paintArea, mapContent, graphic, layout);
```

Comment:
The timeout is there to make sure that we stop the rendering and thus free resources (cpu and memory) locked by the renderer, which are both quite significant in terms of consumption (a reason why we normally tell people to keep a low value for concurrent wms requests in the flow controller).

About painting a large feature, the java2d code needs to go over a PathIterator to do the painting, so we could make it so that the path iterators returned by our custom Shape objects (LiteShape) start throwing exceptions when the stop rendering flag is raised. That happens in the the painting thread, which is a secondary one, so we could also just check if the thread has been interrupted, just not sure how much overhead that will cause (might be very significant, we are talking about millions of extra checks, one per coordinate pair... maybe we could make it so that the check is done only once every 1000 coordinates, or something like that).

But that is just one case. You could be waiting on the database to finish executing your request (see http://blog.grovehillsoftware.com/2009/08/interruptible-jdbc-statements.html for an approach), or you could be waiting for jai to render an image (tile cache poisoning could be a way there, we could build a wrapper around the tile cache that checks if the current thread has been interrupted), or we could be encoding the output, e.g.,  a large png output takes several seconds to perform the encoding, even if the source image is already fully computed (not the jai streaming case you get with direct raster renderer), which might be addressed via output stream poisoning instead. So... as you can see there is a list of things that would need to be done to get the system more or less behave like you want... not a small amount of work, but a "right" solution. The approach you're proposing is instead going to give the illusion of timeout, but cause more problems down the line.